otlp_json2/__init__.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
+
+import json
+from collections.abc import Mapping, Sequence
+from typing import Any, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+    from opentelemetry.trace import Link
+    from opentelemetry._logs import LogRecord
+    from opentelemetry.sdk.trace import ReadableSpan, Event
+    from opentelemetry.sdk.resources import Resource
+    from opentelemetry.sdk.util.instrumentation import InstrumentationScope
+    from opentelemetry.trace.status import Status
+
+    _LEAF_VALUE: TypeAlias = "str | int | float | bool"  # TODO: confirm
+    _VALUE: TypeAlias = "_LEAF_VALUE | Sequence[_LEAF_VALUE]"
+
+
+CONTENT_TYPE = "application/json"
+
+
+_VALUE_TYPES = {
+    # NOTE: order matters, for isinstance(True, int).
+    bool: ("boolValue", bool),
+    int: ("intValue", str),
+    float: ("doubleValue", float),
+    bytes: ("bytesValue", bytes),
+    str: ("stringValue", str),
+    Sequence: (
+        "arrayValue",
+        lambda value: {"values": [_value(e) for e in _homogeneous_array(value)]},
+    ),
+    Mapping: (
+        "kvlistValue",
+        lambda value: {"values": [{k: _value(v) for k, v in value.items()}]},
+    ),
+}
+
+
+def encode_spans(spans: Sequence[ReadableSpan]) -> bytes:
+    spans = sorted(spans, key=lambda s: (id(s.resource), id(s.instrumentation_scope)))
+    rv = {"resourceSpans": []}
+    last_rs = last_is = None
+    for span in spans:
+        assert span.resource
+        assert span.instrumentation_scope
+        if span.resource is not last_rs:
+            last_rs = span.resource
+            last_is = None
+            rv["resourceSpans"].append(
+                {
+                    "resource": _resource(span.resource),
+                    "scopeSpans": [],
+                }
+            )
+        if span.instrumentation_scope is not last_is:
+            last_is = span.instrumentation_scope
+            rv["resourceSpans"][-1]["scopeSpans"].append(
+                {
+                    "scope": _scope(span.instrumentation_scope),
+                    "spans": [],
+                }
+            )
+        rv["resourceSpans"][-1]["scopeSpans"][-1]["spans"].append(_span(span))
+    return json.dumps(rv, separators=(",", ":")).encode("utf-8")
+
+
+def _resource(resource: Resource):
+    # TODO: add schema_url once that lands in opentelemetry-sdk
+    return _attributes(resource)
+
+
+def _attributes(
+    thing: Resource | InstrumentationScope | ReadableSpan | Event | Link | LogRecord,
+) -> dict[str, Any]:
+    rv = {"attributes": [], "dropped_attributes_count": 0}
```

Comment:
Again, I think it'd be clearer/simpler to build the dict at the end, and work with simple Python objects here. And deleting from the dict you've already put them in seems a bit weird, rather than only inserting them if needed. Something like:

```python
attrs = []
for k, v in thing.attributes.items():
    try:
        attrs.append({"key": k, "value": _value(v)})
    except ValueError:
        pass

result = {}
if attrs:
    result["attributes"] = attrs

dropped = len(thing.attributes) - len(attrs)
if dropped:
    result["dropped_attributes_count"] = dropped
```